gadget/mkfs.go

```diff
@@ -0,0 +1,121 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package gadget
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+// MkfsExt4 creates an EXT4 filesystem in given image file, with an optional
+// filesystem label, and populates it with the contents of provided root
+// directory.
+func MkfsExt4(img, label, contentsRootDir string) error {
+	// taken from ubuntu-image
+	// TODO: support e2fsprogs 1.42 without -d in Ubuntu 16.04
+	mkfsArgs := []string{
+		"mkfs.ext4",
+		// default usage type
+		"-T", "default",
+		// disable metadata checksum
+		// XXX: why?
+		"-O", "-metadata_csum",
+		// allow uninitialized block groups
+		"-O", "uninit_bg",
+		// mkfs.ext4 can populate the filesystem with contents of given
+		// root directory
+		"-d", contentsRootDir,
+	}
+	if label != "" {
+		mkfsArgs = append(mkfsArgs, "-L", label)
+	}
+	mkfsArgs = append(mkfsArgs, img)
+	cmd := exec.Command("fakeroot", mkfsArgs...)
```

Comment:
Maybe a comment why we need fakeroot here? Also we probably don't want/need it if we are already root. I'm mostly thinking of an early boot context (initramfs) in which we *may* need to create a filesystem but have only a very limited selection of tools available. This may not be a concern in practise because it looks like we will always have a base snap available in recover mode but being defensive won't hurt I guess.